pkg/pod/status.go

```diff
@@ -314,6 +317,8 @@ func updateIncompleteTaskRunStatus(trs *v1beta1.TaskRunStatus, pod *corev1.Pod)
 			markStatusRunning(trs, ReasonExceededNodeResources, "TaskRun Pod exceeded available resources")
 		case isPodHitConfigError(pod):
 			markStatusFailure(trs, ReasonCreateContainerConfigError, "Failed to create pod due to config error")
+		case isPullImageError(pod):
```

Comment:
Wondering if it would make sense to mark the taskrun status failed for some of these errors? e.g. "invalidImageName" is unlikely to succeed or be retryable

Reply:
Yes, I had thought about this, I tried generating an invalidImageName error for the pod and trying to check the pod's phase. I found that the phase is Pending and not Failed.

<img width="1021" alt="image" src="https://user-images.githubusercontent.com/23007448/169073267-9486ad3f-a6c9-4ca6-b5d5-67ecca104bda.png">


I think this is because of the [following reason](https://github.com/tektoncd/pipeline/issues/4802#issue-1217467406).  
"This happens because it a fixable error from the Pod standpoint (you can mutate the image of a container in a Pod), and thus doesn't seem to be considered as a "terminal" error."

So I thought maybe we could borrow the status of the pod and not mark the taskrun failed?

Reply:
that could definitely work, and I'm happy to approve the PR as is since that's a reasonable route, however the user isn't really meant to interact with the taskrun pod, i.e. the taskrun controller would be responsible for updating the pod and it wouldn't know how to do so here, so I think it's most reasonable to mark as failed.
Curious what others think about this

Reply:
I thought about this a bit more:

- If we mark the TaskRun as pending still, it's not a great user experience because the user isn't intended to interact with the pod and modify the image name, and it'll eventually fail due to timeout (as I mentioned previously)
- However, if we mark the TaskRun as failed, the pod will still be running, and if in theory someone edited the image, the pod could run to completion. If that happened, we wouldn't want the TaskRun to display as failed.

I think the best approach for right now is to mark the TaskRun as pending, preserving existing behavior. Then, we should add functionality that cancels the TaskRun when the pod encounters an error such as InvalidImageName (not as part of this PR-- can just create an issue to track). I think someone is implementing this functionality to cancel pods when the TaskRuns time out (#4618) which could probably be reused here.

What do you think? If you agree, happy to LGTM.

Reply:
Yes! I agree with you.  👍

Reply:
I have started to see this issue a lot on my clusters, where InvalidImageName end up having the PR running forever (or unti default  timeout of 60mn). I think it would be a better user experience if the taskruns would get canceled when encountering those.

Reply:
@chmouel would you mind creating an issue to track this?

Reply:
yep here: https://github.com/tektoncd/pipeline/issues/6105

Reply:
awesome thank you!

Reply:
Uhm, I don't think I would like to support in Tekton a case where the users edit the underlying `Pod` to fix an `ErrInvalidImageName` error.  I would consider that as a permanent error from Tekton perspective. 
However, I do agree that we should not let a `Pod` running when the `TaskRun` is marked as failed. 
If k8s behaviour is to wait even in case of invalid image name, we would then need to stop the pod somehow because we could fail the `TaskRun`.